### PR TITLE
PTCD-869 Refactor and Remove Methods From IApprenticeshipService

### PR DESCRIPTION
**Pull request by jameslawesynergy**
Partially complete while I figure out how to do the remaining methods.

- Remove unused GetStandardByCode, GetFrameworkByCode, GetApprenticeshipDashboardCounts and GetTotalLiveApprenticeships
- Refactored and removed **StandardsAndFrameworksSearch**
- Refactored and removed **GetApprenticeshipByIdAsync**
- Refactored and removed **GetApprenticeshipByUKPRN**
- Refactored and removed **AddApprenticeship**
- Refactored and removed **DeleteBulkUploadApprenticeships**

### CODE REVIEW

File: src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/QueryHandlers/DeleteApprenticeshipHandler.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        public async Task<OneOf<NotFound, Success>> Execute(DocumentClient client, Configuration configuration, Queries.DeleteApprenticeship request)
+        {
+            var documentUri = UriFactory.CreateDocumentUri(
+                configuration.DatabaseId,
+                configuration.ApprenticeshipCollectionName,
+                request.ApprenticeshipId.ToString());
+
+            try
+            {
+                var result = await client.DeleteDocumentAsync(documentUri, new RequestOptions { PartitionKey = new PartitionKey(request.ProviderUkprn) });
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/QueryHandlers/DeleteApprenticeshipHandler.cs:21]: I know we're not supposed to do hard deletes, but https://github.com/SkillsFundingAgency/dfc-providerportal-apprenticeships/blob/main/src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs#L299

File: src/Dfc.CourseDirectory.Web/Controllers/ProviderApprenticeships/ProviderApprenticeshipsController.cs [hunk 2]
```diff
@@ -67,62 +35,65 @@ public async Task<IActionResult> Index(Guid? apprenticeshipId, string message)
...
-                if (string.IsNullOrWhiteSpace(requestModel.SearchTerm))
-                {
-                    model.Items = liveApprenticeships.ToList();
-                }
-                else
-                {
-                    var searchTermWords = requestModel.SearchTerm.Split(' ', StringSplitOptions.RemoveEmptyEntries);
+            var model = new ProviderApprenticeshipsSearchResultModel();
+
+            if (string.IsNullOrWhiteSpace(requestModel.SearchTerm))
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Web/Controllers/ProviderApprenticeships/ProviderApprenticeshipsController.cs:78]: I didn't really want to re-implement search, but it was here... https://github.com/SkillsFundingAgency/dfc-providerportal-apprenticeships/blob/main/src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs#L170-L222

File: src/Dfc.CourseDirectory.Web/Controllers/ApprenticeshipsController.cs [hunk 3]
```diff
@@ -74,54 +83,46 @@ public async Task<IActionResult> Search([FromQuery] SearchRequestModel requestMo
...
-
-            if (!string.IsNullOrEmpty(requestModel.SearchTerm))
+            if (string.IsNullOrWhiteSpace(requestModel?.SearchTerm))
             {
-                var result = await _apprenticeshipService.StandardsAndFrameworksSearch(requestModel.SearchTerm, UKPRN.Value);
+                return ViewComponent(nameof(ApprenticeshipSearchResult), new ApprenticeshipsSearchResultModel());
+            }
 
-                var listOfApprenticeships = new List<ApprenticeShipsSearchResultItemModel>();
+            var searchTermWords = GetSearchWords(requestModel.SearchTerm);
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Web/Controllers/ApprenticeshipsController.cs:91]: I didn't really want to re-implement search, but it was here... https://github.com/SkillsFundingAgency/dfc-providerportal-apprenticeships/blob/main/src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs#L170-L222

File: src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs [hunk 4]
```diff
@@ -297,16 +297,19 @@ public async Task<IActionResult> DeleteFile(DeleteFileViewModel model)
...
+                Predicate = a =>
+                    a.ProviderUKPRN == UKPRN
+                    && (a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadPending || a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadReadyToGoLive)
+            });
+
+            foreach (var apprenticeship in apprenticeships.Values)
             {
-                return RedirectToAction("Index", "Home", new { errmsg = "Delete All Bulk Uploaded Apprenticeships Error" });
+                await _cosmosDbQueryDispatcher.ExecuteQuery(new DeleteApprenticeship { ApprenticeshipId = apprenticeship.Id, ProviderUkprn = apprenticeship.ProviderUKPRN });
             }
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs:310]: Does this need to include the functionality below too?
https://github.com/SkillsFundingAgency/dfc-providerportal-apprenticeships/blob/main/src/Dfc.ProviderPortal.Apprenticeships/Functions/DeleteBulkUploadApprenticeships.cs#L40-L44

File: src/Dfc.CourseDirectory.Web/Controllers/PublishApprenticeships/PublishApprenticeshipsController.cs [hunk 5]
```diff
@@ -19,43 +20,52 @@ namespace Dfc.CourseDirectory.Web.Controllers.PublishApprenticeships
...
+                    a.ProviderUKPRN == UKPRN
+                    && (a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadPending || a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadReadyToGoLive)
+            });
+
+            var bulkUploadPendingApprenticeships = apprenticeships.Values.Where(a => a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadPending).Select(Apprenticeship.FromCosmosDbModel).ToArray();
+            var bulkUploadReadyToGoLiveApprenticeships = apprenticeships.Values.Where(a => a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadReadyToGoLive).Select(Apprenticeship.FromCosmosDbModel).ToArray();
 
-            if (vm.AreAllReadyToBePublished)
+            vm.ListOfApprenticeships = GetErrorMessages(bulkUploadPendingApprenticeships);
+
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Web/Controllers/PublishApprenticeships/PublishApprenticeshipsController.cs:60]: Does the `AreAllReadyToBePublished` property need to be populated on the VM?

File: src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs (hunk 4 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs:310]: I did contemplate that, but I tried to rationalise why that was in there in the first place.

As far as I'm aware, we don't ever set the `Apprenticeship.RecordStatus` to `MigrationPending` or `MigrationReadyToGoLive`, do we?

This seems to me like a hack to clean up some apprenticeships with legacy statuses.

File: src/Dfc.CourseDirectory.Web/Controllers/PublishApprenticeships/PublishApprenticeshipsController.cs (hunk 5 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Web/Controllers/PublishApprenticeships/PublishApprenticeshipsController.cs:60]: Good spot

File: src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs (hunk 4 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs:310]: We don't set any new records to that status but we may still have some old ones around.

Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs:310]: Should cancelling an apprenticeships bulk upload really trigger that process? If the intention is to only clean up old records and nothing to do with the bulk upload process, then I vote to omit these behaviours.

Comment (gunndabad) [reply src/Dfc.CourseDirectory.Web/Controllers/BulkUploadApprenticeshipsController.cs:310]: It doesn't make sense to me (I can understand it when publishing a bulk upload but not when deleting pending records). If you're happy to take the behaviour change then I'm fine with it.

**Review by gunndabad — APPROVED** [review]

**Pull request merged**